### ISSUE TITLE
Proposal: Multiple instances of integrations

### ISSUE THREAD

**Issue by rfratto** [grafana/agent#1059]
(This is a more design-y proposal for #350)

It is not currently possible to run multiple instances of integrations. This is a proposal to enable this to happen. 

## Configuration 

Currently, each integration may be defined in the YAML exactly once, via a `<integration name>` key inside of the `integrations` block. 

I propose adding a new `<integrations name>_configs` key for each integration, which holds an array of integration configs. Elements in this array will default to `enabled: true`, a change from the previous key. 

The old style of defining integrations will be _marked as deprecated and eventually removed_. Config files will fail to load if both the old style and new style are used simultaneously for a given integration. 

Old YAML:

```yaml
integrations:
  memcached_exporter:
    enabled: true
    memcached_address: localhost:11211 
```

New YAML:

```yaml
integrations:
  memcached_exporter_configs:
    - memcached_address: localhost:11211     
    - memcached_address: localhost:11212     
    - memcached_address: localhost:11213 
```

## HTTP Path 

The HTTP path used for integration metrics today must change. Today, each integration a unique HTTP endpoint at `/integrations/<integration name>/metrics`. Individual instances of the same integration would need to be given different endpoints so they can be scraped as separate targets. 

I propose using the instance key from #1033 as part of the path name for the HTTP endpoint. This key would need to be unique (it's currently not validated, IIRC) for each integration instance. The new path for an integration would be collectable via `/integration/<integration name>/<instance key>/metrics`.  

For example, this config:

```yaml
integrations:
  memcached_exporter_configs:
    - memcached_address: localhost:11211     
    - memcached_address: localhost:11212     
    - memcached_address: localhost:11213 
```

Will generate three endpoints (using URL-encoding for the instance key where required):

* `/integration/memcached_exporter/localhost%3A11211/metrics` 
* `/integration/memcached_exporter/localhost%3A11212/metrics` 
* `/integration/memcached_exporter/localhost%3A11213/metrics` 

This new HTTP path is only used for the `_configs`-style integrations, so old paths will continue to work. Once a user migrates to the _configs-style format, they must update any scrape jobs they have that use that old path. 

## Integration Instance Listing

Navigating to `/integrations/<integration name>` should show a listing of all metrics endpoints as HTML to make it easier for users to manually discover and configure scrape jobs for these integrations.

**Comment by rfratto** [comment]
> Will generate three endpoints (using URL-encoding for the instance key where required):

The URL-encoding is ugly, but a user could override the instance key to help with this:

```yaml
integrations:
  memcached_exporter_configs:
  # defines /integration/memcached_exporter/server-a/metrics
  - memcached_address: localhost:11211 
    instance: server-a 
```

**Comment by rfratto** [comment]
Also, since there's been talk about conf.d-style loading in #868, I think the array approach here would help a lot: an integrations file could contain something like

```yaml
memcached_exporter_configs:
- memcached_address: localhost:11211 
```

and easily get concatenated into the full set of defined integrations without having to worry about order/merge rules.

**Comment by gaantunes** [comment]
Sounds good to me, liked the idea of having a page to show the endpoints, and being able to override the key with the `instance `label.

Are we going to limit the number of instances per agent? I wonder if the Agent can keep an indefinite number of integration instances, and if not what is this limit, which changes case by case.

**Comment by rfratto** [comment]
Thinking about this more, we might not want to deprecate the old config right away. Maybe this should be something we consider experimental for a release or two?

> Are we going to limit the number of instances per agent? I wonder if the Agent can keep an indefinite number of integration instances, and if not what is this limit, which changes case by case.

I don't think we would, a user should be able to configure as many as they need for their environment, but the limitation here would be how much memory each consumes over time.

**Comment by hjet** [comment]
Will we support sharing config parameters across integration instances? Or would this be necessary:

```
integrations:
  memcached_exporter_configs:
    - memcached_address: localhost:11211
       relabel_configs:
       - source_labels: [foo]
          target_label: bar
          replacement: baz
    - memcached_address: localhost:11212
       relabel_configs:
       - source_labels: [foo]
          target_label: bar
          replacement: baz
    - memcached_address: localhost:11213
       relabel_configs:
       - source_labels: [foo]
          target_label: bar
          replacement: baz
```

**Comment by rfratto** [comment]
@hjet yeah, you'd have to have relabel configs set up for each. YAML has tricks (like anchors) for this, but I don't think we'd support something natively in the schema to reduce duplication.

**Comment by rfratto** [comment]
Example of using anchors:

```yaml
integrations:
  memcached_exporter_configs:
    - memcached_address: localhost:11211
      relabel_configs: &memcached_relabel_configs
       - source_labels: [foo]
         target_label: bar
         replacement: baz
    - memcached_address: localhost:11212
      relabel_configs: *memcached_relabel_configs
    - memcached_address: localhost:11213
      relabel_configs: *memcached_relabel_configs
```

**Comment by rgeyer** [comment]
A bit late to the party, but this all looks good to me.

What would be the difficulty of supporting anchors? While it's an advanced feature, I would think it might actually be useful.

**Comment by rfratto** [comment]
> What would be the difficulty of supporting anchors? While it's an advanced feature, I would think it might actually be useful.

It should be supported already by the [YAML parser we use](https://pkg.go.dev/gopkg.in/yaml.v2#section-readme):

> The yaml package supports most of YAML 1.1 and 1.2, including support for anchors, tags, map merging, etc.

**Comment by rfratto** [comment]
Closed by #1196.
